### PR TITLE
DSPDC-130 Add first draft at an agent template

### PR DESCRIPTION
**Pull request by danxmoran**
Hard to tell how well it actually works without a concrete implementation, but I wanted to get what I've written so far checked in.

### CODE REVIEW

**Comment by danxmoran** [comment]
Looks like the branch build ran on the Jenkins master, which doesn't have Docker. I'll see if I can figure out the right restriction to list in the `Jenkinsfile`.

File: agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+     * Kafka Streams supports deserializers for higher-level types than `Array[Byte]`,
+     * but the interface for those serializers doesn't allow for transforming decoding
+     * errors into output messages.
+     *
+     * Since we rely on the agent reporting a status for every message it processes,
+     * we use the lowest-level deserializer at the stream entrypoint, then roll our own
+     * decoding / error handling logic.
+     */
+    builder
+      .stream[String, Array[Byte]](queue.requestTopic)
```
Comment (milkbeard) [inline agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala]: if  a topic is requested it is presumably not available on the queue any more. couple questions which may seem obvious.
1) What prevents multiple agents picking up the same topic?
2) What ensures that if the agent working on a topic fails we don't lose that transfer? (not just errors I'm thinking more  the vm running the agent died or was shut down)

File: common/src/main/scala/org/broadinstitute/transporter/transfer/TransferStatus.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  case object Success extends TransferStatus
+
+  /**
+    * Signal for transfers that failed with a transient error and
+    * are likely to succeed if resubmitted.
+    *
+    * Whether or not a failure should be considered retriable is
+    * left up to the individual agent implementations.
+    */
+  case object RetriableFailure extends TransferStatus
```
Comment (milkbeard) [inline common/src/main/scala/org/broadinstitute/transporter/transfer/TransferStatus.scala]: retriable or retryable .. both seem equally awkward.

File: agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    /*
+     * Kafka Streams supports deserializers for higher-level types than `Array[Byte]`,
+     * but the interface for those serializers doesn't allow for transforming decoding
+     * errors into output messages.
+     *
+     * Since we rely on the agent reporting a status for every message it processes,
+     * we use the lowest-level deserializer at the stream entrypoint, then roll our own
+     * decoding / error handling logic.
+     */
+    builder
```
Comment (jacarey) [inline agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala]: What prevents multiple agents picking up the same topic?
What ensures that if the agent working on a topic fails we don't lose that transfer? (not just errors I'm thinking more the vm running the agent died or was shut down)

File: common/src/main/scala/org/broadinstitute/transporter/transfer/TransferStatus.scala (hunk 2 above)
Comment (jacarey) [inline common/src/main/scala/org/broadinstitute/transporter/transfer/TransferStatus.scala]: retriable or retryable .. both seem equally awkward.

File: agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala (hunk 3 above)
Comment (danxmoran) [reply agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala]: > What prevents multiple agents picking up the same topic?

For now, nothing. Longer-term I think ACLs would help with this, though I'm not sure if they'd cover it entirely. If it's possible to set ACLs based on consumer group ID we could potentially store an ID whitelist in the `queues` table, and have the manager flip switches when creating new topics.

> What ensures that if the agent working on a topic fails we don't lose that transfer? (not just errors I'm thinking more the vm running the agent died or was shut down)

Kafka Streams claims that it's transactional, and doesn't ack the input topic until a response has been pushed onto the output topic. I'll investigate how much work it'd be to integration-test that.

File: common/src/main/scala/org/broadinstitute/transporter/transfer/TransferStatus.scala (hunk 2 above)
Comment (danxmoran) [reply common/src/main/scala/org/broadinstitute/transporter/transfer/TransferStatus.scala]: Maybe just `TransientFailure` or `TemporaryFailure` instead?

Comment (jacarey) [reply common/src/main/scala/org/broadinstitute/transporter/transfer/TransferStatus.scala]: Sounds fancier.. i like it.

File: agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala (hunk 3 above)
Comment (jacarey) [reply agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala]: For 1. .. Does it make sense for the agents to be able to talk to each other and check ?  

Agent 1 grabs message A.
Agent 2 grabs message A.

Agent 1 says "Do any other agents have A?"
Agent 2 says "I do"
Agent 1 says "Ok go ahead and grab a different message.. I'll handle A.
Agent 2 says "OK"

Agent 2 grabs message B.

Comment (danxmoran) [reply agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala]: Ah, I see what you mean.

Kafka handles this sort of handshake to make sure only one consumer per group ever sees each message. For Streams apps the group is defined by the configured application ID, so as long as we deploy agent replicas with identical values for that field we should be good. I worry that adding cross-agent communication replicates a bunch of the functionality Kafka is meant to provide for us.

Comment (jacarey) [reply agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala]: lovely 👍

**Pull request merged**